### ISSUE TITLE
Sidekiq Crashes

### ISSUE THREAD

**Issue by joshmyers** [mperham/sidekiq#1987]
Gemfile:

```
gem 'sidekiq', '~> 3.2.5'
gem 'sidekiq-cron', '~> 0.2.0'
gem 'sidekiq-failures'
gem 'sidekiq-statsd'
```

config/initializers/sidekiq.rb:

``` ruby
Sidekiq.configure_server do |config|
  config.redis = { :url => ENV['REDIS_URL'], :namespace => 'sidekiq' }

  schedule_file = "config/schedule.yml"
  if File.exists?(schedule_file)
    Sidekiq::Cron::Job.load_from_hash YAML.load_file(schedule_file)
  end

  config.server_middleware do |chain|
    chain.add Sidekiq::Statsd::ServerMiddleware,
    env: 'foo',
    prefix: "bar-#{Node.environment}.app.workers",
    host: ENV['STATSD_ADDR'].split(':').first,
    port: 8125
  end
end

Sidekiq.configure_client do |config|
  config.redis = { :url => ENV['REDIS_URL'], :namespace => 'sidekiq' }
end

Sidekiq.default_worker_options = { 'backtrace' => true , 'retry' => true }

module Sidekiq
  module ExceptionHandler
    def handle_exception(ex, msg)
      # Only send exceptions to wherever we want to send it after 3 tries. If
      # a third party (Instagram, S3), fails for some reason and we retry it
      # where it succeeds, we don't need that to clog up our inboxes. We'll get
      # notified after 3 retries, which means there's something really wrong.
      return unless msg['retry_count'].to_i >= 3

      Sidekiq.logger.warn msg
      Sidekiq.logger.warn ex
      Sidekiq.logger.warn ex.backtrace.join("\n")

      send_to_airbrake(msg, ex) if defined?(::Airbrake)
      send_to_honeybadger(msg, ex) if defined?(::Honeybadger)
      send_to_exceptional(msg, ex) if defined?(::Exceptional)
      send_to_exception_notifier(msg, ex) if defined?(::ExceptionNotifier)
    end
  end
end
```

We are using Runit to manage our Sidekiq processes with the following config file:

```

---
:verbose: true
:pidfile: sidekiq-1
:logfile: /opt/foo/bar/shared/sidekiq/sticky9-1.log
:concurrency: 25
:processes: 1
:timeout: 30
:queues:
  - [high, 10]
  - [default, 5]
  - [low, 2]
```

Each "worker" box has 3 Sidekiq processes running. 2 as above ( concurrency: 25 ) pulling from the default queue and another Sidekiq process with concurrency: 4, working on another queue. So maximum concurrency of 54, spread across 3 processes. Each worker box is an m3.large.

We are seeing a few exceptions thrown and Sidekiq crashes in logs, which is to be expected. Sidekiq seems to recover OK from most of these crashes _except_ for the following:

```
2014-10-06T10:41:07.972Z 867 TID-191lrg ERROR: Sidekiq::Processor crashed!
Celluloid::TimeoutError: linking timeout of 5 seconds exceeded
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:217:in `block (2 levels) in linking_request'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:203:in `loop'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:203:in `block in linking_request'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks.rb:109:in `exclusive'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid.rb:431:in `exclusive'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:198:in `linking_request'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:90:in `monitor'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:101:in `link'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid.rb:199:in `new_link'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/manager.rb:100:in `block in processor_died'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/util.rb:15:in `watchdog'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/sidekiq-3.2.5/lib/sidekiq/manager.rb:94:in `processor_died'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:357:in `handle_exit_event'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:340:in `block in handle_system_event'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:416:in `block in task'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks.rb:55:in `block in initialize'
        /opt/foo/bar/shared/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks/task_fiber.rb:13:in `block in create'
```

We are using RDS MySQL and ElastiCache for Redis ( AWS all the things ). I've checked Redis latency using --latency-history and it doesn't seem _ that_ bad.

```
min: 0, max: 82, avg: 0.79 (1320 samples) -- 15.00 seconds range
min: 0, max: 91, avg: 0.73 (1278 samples) -- 15.00 seconds range
min: 0, max: 18, avg: 0.62 (1339 samples) -- 15.00 seconds range
min: 0, max: 87, avg: 0.86 (1264 samples) -- 15.01 seconds range
min: 0, max: 99, avg: 1.04 (1292 samples) -- 15.00 seconds range
min: 0, max: 91, avg: 0.80 (1271 samples) -- 15.01 seconds range
min: 0, max: 83, avg: 0.72 (1325 samples) -- 15.01 seconds range
min: 0, max: 87, avg: 0.72 (1279 samples) -- 15.01 seconds range
min: 0, max: 52, avg: 0.61 (1342 samples) -- 15.01 seconds range
min: 0, max: 63, avg: 0.99 (1253 samples) -- 15.00 seconds range
```

When Sidekiq crashes the processes still look to be present, but nothing shows in the UI.

```
pgrep -fl sidekiq
9000 runsv sidekiq-foo-1
9053 svlogd -tt /var/log/sidekiq-foo-1
9127 runsv sidekiq-foo-2
9278 svlogd -tt /var/log/sidekiq-foo-2
9369 runsv sidekiq-foo-printing-1
9590 svlogd -tt /var/log/sidekiq-foo-printing-1
24584 sidekiq 3.2.5 bar [2 of 2 busy]                                                                                                          
24731 sidekiq 3.2.5 bar [24 of 25 busy]                                                                                               
24803 sidekiq 3.2.5 bar [24 of 25 busy] 
```

sidekiq-foo-1 is actually not processing any jobs right now. See below screenshot. It drops off the UI, there should be another process in here.

![Sidekiq](http://s1.postimg.org/akeah4en3/Screenshot_2014_10_06_14_15_25.png)

This is proving to be a real pain in the ass so any help/input would be much appreciated. Please ask if any more info would be helpful..

Thanks.

**Comment by diddeb** [comment]
We're seeing the same problem with Sidekiq Pro on Heroku. It's more frequent when we try with concurrency > ~10.

`DB_POOL` is very generous and the RedisToGo instance has a bunch of available connections. As far as I can see, there's no limits being hit or even remotely touched.

```
app[worker.1]: 2015-01-13T23:09:16.316Z 3 TID-ovmmgtqg4 INFO: Sidekiq Pro 1.9.2, commercially licensed.  Thanks for your support!
app[worker.1]: 2015-01-13T23:09:16.234Z 3 TID-ovmmgtqg4 INFO: Running in ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-linux]
app[worker.1]: 2015-01-13T23:09:16.416Z 3 TID-ovmmgtqg4 INFO: Starting processing, hit Ctrl-C to stop
app[worker.1]: 2015-01-13T23:09:26.127Z 3 TID-ovmms6yyo WARN: task was terminated
app[worker.1]: 2015-01-13T23:09:26.127Z 3 TID-ovmms6yyo WARN: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/tasks/task_fiber.rb:34:in `terminate'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:345:in `each'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:345:in `cleanup'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:329:in `shutdown'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:321:in `handle_crash'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:166:in `rescue in run'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:148:in `run'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:130:in `block in start'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/thread_handle.rb:13:in `block in initialize'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor_system.rb:32:in `block in get_thread'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/internal_pool.rb:130:in `call'
app[worker.1]: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/internal_pool.rb:130:in `block in create'
app[worker.1]: 2015-01-13T23:09:26.127Z 3 TID-ovmmmyojw WARN: Sidekiq died due to the following error, cannot recover, process exiting
app[worker.1]: 2015-01-13T23:09:26.127Z 3 TID-ovmmmyojw WARN: linking timeout of 5 seconds exceeded
app[worker.1]: 2015-01-13T23:09:26.127Z 3 TID-ovmmmyojw WARN: /app/vendor/bundle/ruby/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:206:in `
```

If we lower the concurrency to something closer to two, or three this does not happen, AFAICS.

Any ideas? I would be glad to help out with more details and/or try different configurations.

**Comment by Davidslv** [comment]
@diddeb I would recommend you to open a new ticket as this was closed by @joshmyers

**Comment by diddeb** [comment]
@Davidslv :+1:

**Comment by kremso** [comment]
@mperham Do you have any idea what might be causing this? We are hitting this pretty regularly (several times a day) and cannot afford to reduce concurrency. Can I do anything to help debug this?

celluloid 0.16.0 and sidekiq 3.3.4

**Comment by mperham** [comment]
@kremso I need a backtrace or crash report of some type.

**Comment by kremso** [comment]
````
2015-05-14T13:56:34.566Z 7288 TID-otodhromk WARN: Sidekiq died due to the following error, cannot recover, process exiting
2015-05-14T13:56:34.570Z 7288 TID-otodhromk WARN: linking timeout of 5 seconds exceeded
2015-05-14T13:56:34.570Z 7288 TID-otodhromk WARN: /home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:206:in `block in linking_request'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/tasks.rb:107:in `exclusive'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid.rb:414:in `exclusive'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:177:in `linking_request'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:57:in `monitor'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:68:in `link'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid.rb:179:in `new_link'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/sidekiq-3.3.4/lib/sidekiq/manager.rb:102:in `block in processor_died'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/sidekiq-3.3.4/lib/sidekiq/util.rb:16:in `watchdog'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/sidekiq-3.3.4/lib/sidekiq/manager.rb:96:in `processor_died'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/cell.rb:15:in `block in call'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/cell.rb:71:in `block in task'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:357:in `block in task'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/tasks.rb:57:in `block in initialize'
/home/luigis/deploy/shared/bundle/ruby/2.2.0/gems/celluloid-0.16.0/lib/celluloid/tasks/task_fiber.rb:15:in `block in create'
````

**Comment by mperham** [comment]
That error is always due to inefficient Sidetiq recurrence declarations.  Are you using Sidetiq?

**Comment by kremso** [comment]
No, we are not using sidetiq.

**Comment by mperham** [comment]
Then I don't have any idea what it causing it.  Are you running Sidekiq on a really small instance/machine with little CPU?

**Comment by kremso** [comment]
24 cores, 64Gb RAM with plenty of headroom. Do you suspect this is CPU related?

**Comment by mperham** [comment]
I was suspecting CPU starvation but I guess not.  No clue.  This is a relatively rare issue that is usually due to sidetiq hogging the CPU.

**Comment by kremso** [comment]
I checked CPU usage at the time of today's crash and it looks normal. Thanks for trying to help and please let me know when you have any other idea.

**Comment by kremso** [comment]
We are calling a (sometimes slow) http service from our workers. I noticed that at the time when sidekiq crashes we also observed increased response times from the service to the point that the http requests from workers timeout.

I put up a minimal setup that simulates this and it can reproduce the issue reliably; see here https://gist.github.com/kremso/ffee0659f40ea630825f

It takes around a minute or two, but eventually sidekiq will crash with the linking timeout. I also noticed that with http timeout length <= 2 seconds the issue goes away, with timeout larger than 2 seconds, sidekiq will crash.

@mperham Does this give you any idea what might by going on?

**Comment by mperham** [comment]
That code looks fine so I suspect Patron.  Can you try the same with plain old Net::HTTP?

**Comment by kremso** [comment]
You were right, it does not crash with Net::HTTP (https://gist.github.com/kremso/ffee0659f40ea630825f#file-net_http_worker-rb).

What does it mean? Those patron workers starve CPU for 5 seconds? I didn't see any increased CPU utilization during my tests. In general, what behavior can cause a linking timeout?

**Comment by mperham** [comment]
I don't know, that's a question for the celluloid team.

**Comment by ryansch** [comment]
I'd be interested to know how `excon` performs in that test.  Still trying to find an http backend that just works. :sob:

**Comment by guyisra** [comment]
Is there a way to automatically restart sidekiq in this situation?

**Comment by mperham** [comment]
@guyisra Not that I'm aware of.

**Comment by milgner** [comment]
@guyisra: you can monitor your processes with (eye)[https://github.com/kostya/eye], (god)[http://godrb.com/] or (bluepill)[https://github.com/arya/bluepill]. If you can't prevent the crash, they can at least restart the component when it does.

Imho it's best practice to do this for all your processes, including the web-server.

Additionally you can have them monitor CPU load and/or memory consumption to notify you about undesired behaviour there.

**Comment by kremso** [comment]
@guyisra In this particular case, sidekiq crashes (the process terminates). If you run it from init then you can configure it to start automatically again (like https://github.com/mperham/sidekiq/tree/master/examples/upstart).

**Comment by jaredjenkins** [comment]
FWIW, we saw this issue come up with JRuby running on Java 8. We were able stabilize things with adding this flag to our JRuby boot options: `-Xcompile.invokedynamic=false`. This option was previously false by default in Java 7 but defaults to true in Java 8. Supposedly it can boost [JRuby performance](https://github.com/jruby/jruby/wiki/PerformanceTuning#enabling-invokedynamic), but as far as I can tell it was causing long pauses trying to spin up new threads. Celluloid will crash if it can't get a new thread within 5 seconds. This happened regardless of JRuby version tried 1.7.12 and 9.0.1.0.
